File: app/routers/booking.py
"""Client-facing booking flow routes."""

from __future__ import annotations

import logging
from datetime import date, datetime, timedelta
from zoneinfo import ZoneInfo

from fastapi import APIRouter, HTTPException, Request, Response
from fastapi.responses import HTMLResponse
from app.templates_env import get_templates
from pydantic import BaseModel, EmailStr
from sqlalchemy import select
from sqlalchemy.orm import selectinload

from app.dependencies import CalendarDep, DbSession, EmailDep, WebhookDep
from app.models.booking import Booking, BookingStatus
from app.models.event_type import EventType
from app.models.user import User
from app.services.auth import generate_token
from app.services.availability import detect_timezone_warning, generate_slots_for_date
from app.services.ics import generate_ics

logger = logging.getLogger(__name__)
router = APIRouter(tags=["booking"])
templates = get_templates()


# ---------------------------------------------------------------------------
# Public booking page — /{slug}
# ---------------------------------------------------------------------------


@router.get("/{slug}", response_class=HTMLResponse)
async def booking_page(slug: str, request: Request, db: DbSession):
    result = await db.execute(
        select(User)
        .where(User.slug == slug, User.is_active == True)
        .options(selectinload(User.event_types), selectinload(User.availability_rules))
    )
    user = result.scalar_one_or_none()
    if user is None:
        raise HTTPException(status_code=404, detail="Booking page not found")

    active_event_types = [et for et in user.event_types if et.is_active]
    return templates.TemplateResponse(
        "pages/booking_page.html",
        {"request": request, "user": user, "event_types": active_event_types},
    )


# ---------------------------------------------------------------------------
# Slot availability API
# ---------------------------------------------------------------------------


@router.get("/{slug}/{event_type_slug}/slots")
async def get_slots(
    slug: str,
    event_type_slug: str,
    target_date: date,
    client_timezone: str = "UTC",
    db: DbSession = None,
    calendar: CalendarDep = None,
):
    """Return available slots for a given date. Called via HTMX."""
    user, event_type = await _load_user_and_event_type(slug, event_type_slug, db)

    # Compute the full busy window for the day
    tz = ZoneInfo(user.timezone)
    day_start = datetime.combine(target_date, datetime.min.time(), tzinfo=tz)
    day_end = day_start + timedelta(days=1)
    day_start_utc = day_start.astimezone(ZoneInfo("UTC"))
    day_end_utc = day_end.astimezone(ZoneInfo("UTC"))

    busy = await calendar.get_free_busy(user.id, day_start_utc, day_end_utc)

    # Also treat confirmed bookings in DB as busy
    db_busy = await _get_db_busy(user.id, day_start_utc, day_end_utc, db)
    all_busy = busy + db_busy

    slots = generate_slots_for_date(
        target_date,
        user.availability_rules,
        all_busy,
        event_type,
        user.timezone,
    )

    # Format slots in both UTC and client timezone
    client_tz = ZoneInfo(client_timezone)
    slot_data = []
    for slot_utc in slots:
        slot_local = slot_utc.astimezone(client_tz)
        end_utc = slot_utc + timedelta(minutes=event_type.duration_minutes)
        warning = detect_timezone_warning(slot_utc, end_utc, user.timezone, client_timezone)
        slot_data.append({
            "utc": slot_utc.isoformat(),
            "local": slot_local.strftime("%I:%M %p"),
            "warning": warning,
        })

    return {"date": target_date.isoformat(), "slots": slot_data}


# ---------------------------------------------------------------------------
# Booking submission
# ---------------------------------------------------------------------------


class BookingRequest(BaseModel):
    slot_utc: datetime
    client_name: str
    client_email: EmailStr
    client_timezone: str = "UTC"
    intake_answers: dict = {}


@router.post("/{slug}/{event_type_slug}/book")
async def create_booking(
    slug: str,
    event_type_slug: str,
    data: BookingRequest,
    db: DbSession,
    calendar: CalendarDep,
    email: EmailDep,
    webhook: WebhookDep,
    request: Request,
):
    user, event_type = await _load_user_and_event_type(slug, event_type_slug, db)

    start_utc = data.slot_utc.replace(tzinfo=ZoneInfo("UTC")) if data.slot_utc.tzinfo is None else data.slot_utc
    end_utc = start_utc + timedelta(minutes=event_type.duration_minutes)

    # Double-check slot is still available
    day_start = start_utc.replace(hour=0, minute=0, second=0, microsecond=0)
    day_end = day_start + timedelta(days=1)
    busy = await calendar.get_free_busy(user.id, day_start, day_end)
    db_busy = await _get_db_busy(user.id, day_start, day_end, db)

    from app.services.availability import _overlaps_any_busy
    slot_end_with_buffer = end_utc + timedelta(minutes=event_type.buffer_minutes)
    if _overlaps_any_busy(start_utc, slot_end_with_buffer, busy + db_busy):
        raise HTTPException(status_code=409, detail="Slot no longer available")

    # Create the booking
    booking = Booking(
        event_type_id=event_type.id,
        client_email=data.client_email,
        client_name=data.client_name,
        client_timezone=data.client_timezone,
        client_data=data.intake_answers,
        start_at=start_utc,
        end_at=end_utc,
        status=BookingStatus.confirmed,
        cancel_token=generate_token(),
        reschedule_token=generate_token(),
    )
    db.add(booking)
    await db.flush()

    # Create calendar event
    description = f"Booking with {data.client_name}\n\n"
    if data.intake_answers:
        for k, v in data.intake_answers.items():
            description += f"{k}: {v}\n"

    try:
        event_id = await calendar.create_event(
            user.id,
            f"{event_type.title} — {data.client_name}",
            start_utc,
            end_utc,
            description=description,
            attendee_email=data.client_email,
            create_meet_link=True,
        )
        booking.google_event_id = event_id
    except Exception as exc:
        logger.warning("Calendar event creation failed: %s", exc)

    await db.commit()
    await db.refresh(booking)

    base_url = str(request.base_url).rstrip("/")
    cancel_url = f"{base_url}/bookings/{booking.cancel_token}/cancel"
    reschedule_url = f"{base_url}/bookings/{booking.reschedule_token}/reschedule"

    # Generate ICS
    ics = generate_ics(
        uid=f"booking-{booking.id}@bisque-booking",
        title=f"{event_type.title} with {user.name}",
        start=start_utc,
        end=end_utc,
        organizer_email=user.email,
        organizer_name=user.name,
        attendee_email=data.client_email,
        attendee_name=data.client_name,
        description=description,
        location=event_type.video_link or event_type.location or "",
    )

    # Send confirmation emails
    try:
        await email.send_confirmation_to_client(
            booking_id=booking.id,
            client_email=data.client_email,
            client_name=data.client_name,
            consultant_name=user.name,
            start=start_utc,
            end=end_utc,
            client_timezone=data.client_timezone,
            cancel_url=cancel_url,
            reschedule_url=reschedule_url,
            video_link=event_type.video_link,
            ics_content=ics,
        )
        await email.send_confirmation_to_consultant(
            booking_id=booking.id,
            consultant_email=user.email,
            consultant_name=user.name,
            client_name=data.client_name,
            client_email=data.client_email,
            client_data=data.intake_answers,
            start=start_utc,
            end=end_utc,
            consultant_timezone=user.timezone,
            cancel_url=cancel_url,
            ics_content=ics,
        )
    except Exception as exc:
        logger.error("Email sending failed: %s", exc)

    # Fire webhook
    await webhook.dispatch("booking.created", {
        "booking_id": booking.id,
        "client_email": booking.client_email,
        "start_at": start_utc.isoformat(),
        "end_at": end_utc.isoformat(),
        "event_type": event_type.slug,
        "consultant_slug": user.slug,
    })

    return {
        "booking_id": booking.id,
        "start_at": start_utc.isoformat(),
        "end_at": end_utc.isoformat(),
        "cancel_url": cancel_url,
        "reschedule_url": reschedule_url,
    }


# ---------------------------------------------------------------------------
# Cancel / reschedule
# ---------------------------------------------------------------------------


@router.get("/bookings/{token}/cancel")
async def cancel_booking(token: str, db: DbSession, email: EmailDep, calendar: CalendarDep, request: Request):
    result = await db.execute(
        select(Booking)
        .where(Booking.cancel_token == token)
        .options(selectinload(Booking.event_type).selectinload(EventType.user))
    )
    booking = result.scalar_one_or_none()
    if booking is None or booking.status == BookingStatus.cancelled:
        raise HTTPException(status_code=404, detail="Booking not found or already cancelled")

    booking.status = BookingStatus.cancelled
    if booking.google_event_id:
        try:
            await calendar.delete_event(booking.event_type.user_id, booking.google_event_id)
        except Exception as exc:
            logger.warning("Calendar delete failed: %s", exc)

    await db.commit()

    # Send cancellation emails
    user = booking.event_type.user
    try:
        await email.send_cancellation(
            booking.id, booking.client_email, booking.client_name,
            booking.start_at, booking.end_at, booking.client_timezone,
        )
        await email.send_cancellation(
            booking.id, user.email, user.name,
            booking.start_at, booking.end_at, user.timezone,
        )
    except Exception as exc:
        logger.error("Cancellation email failed: %s", exc)

    return {"status": "cancelled"}


@router.get("/bookings/{token}/reschedule")
async def reschedule_redirect(token: str, db: DbSession):
    """Cancel the current booking and redirect to the booking page to pick a new slot."""
    result = await db.execute(
        select(Booking)
        .where(Booking.reschedule_token == token)
        .options(selectinload(Booking.event_type).selectinload(EventType.user))
    )
    booking = result.scalar_one_or_none()
    if booking is None:
        raise HTTPException(status_code=404, detail="Booking not found")

    user = booking.event_type.user
    event_type = booking.event_type
    booking.status = BookingStatus.rescheduled

    await db.commit()
    from fastapi.responses import RedirectResponse
    return RedirectResponse(url=f"/{user.slug}/{event_type.slug}", status_code=303)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


async def _load_user_and_event_type(slug: str, event_type_slug: str, db) -> tuple[User, EventType]:
    result = await db.execute(
        select(User)
        .where(User.slug == slug, User.is_active == True)
        .options(selectinload(User.availability_rules))
    )
    user = result.scalar_one_or_none()
    if user is None:
        raise HTTPException(status_code=404, detail="User not found")

    et_result = await db.execute(
        select(EventType).where(EventType.user_id == user.id, EventType.slug == event_type_slug, EventType.is_active == True)
    )
    event_type = et_result.scalar_one_or_none()
    if event_type is None:
        raise HTTPException(status_code=404, detail="Event type not found")

    return user, event_type


async def _get_db_busy(user_id: int, start: datetime, end: datetime, db) -> list[tuple[datetime, datetime]]:
    """Get confirmed bookings for a user as busy intervals."""
    result = await db.execute(
        select(Booking)
        .join(EventType)
        .where(
            EventType.user_id == user_id,
            Booking.status == BookingStatus.confirmed,
            Booking.start_at < end,
            Booking.end_at > start,
        )
    )
    return [(b.start_at, b.end_at) for b in result.scalars()]

File: ts/src/routes/booking.ts
/**
 * Client-facing booking flow routes.
 * Mirrors app/routers/booking.py.
 *
 * Routes:
 *   GET  /:slug                          → booking page (user profile + event types)
 *   GET  /:slug/:eventTypeSlug/slots     → available slots for a date
 *   POST /:slug/:eventTypeSlug/book      → create a booking
 *   GET  /bookings/:token/cancel         → cancel a booking
 *   GET  /bookings/:token/reschedule     → begin reschedule
 */

import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { z } from "zod";
import { and, eq, gte, lt } from "drizzle-orm";
import {
  getDb,
  users,
  eventTypes,
  availabilityRules,
  bookings,
} from "@/db";
import type { CalendarProvider, EmailProvider, WebhookProvider } from "@/providers/types";
import { NoopCalendarProvider, NoopEmailProvider, NoopWebhookProvider } from "@/providers/noop";
import { generateToken } from "@/auth/token-utils";
import {
  generateSlotsForDate,
  detectTimezoneWarning,
  overlapsAnyBusy,
  type BusyInterval,
  type CalendarDate,
} from "@/slots/engine";
import { generateIcs } from "@/ics";
import { getConfig } from "@/config";

const booking = new Hono();

// ---------------------------------------------------------------------------
// Provider injection (DI via closures — replaced by real providers in prod)
// ---------------------------------------------------------------------------

let _calendarProvider: CalendarProvider = new NoopCalendarProvider();
let _emailProvider: EmailProvider = new NoopEmailProvider();
let _webhookProvider: WebhookProvider = new NoopWebhookProvider();

export function setProviders(opts: {
  calendar?: CalendarProvider;
  email?: EmailProvider;
  webhook?: WebhookProvider;
}): void {
  if (opts.calendar) _calendarProvider = opts.calendar;
  if (opts.email) _emailProvider = opts.email;
  if (opts.webhook) _webhookProvider = opts.webhook;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function first<T>(arr: T[]): T | undefined {
  return arr[0];
}

/** Convert a UTC Date to CalendarDate (year, month, day) in UTC. */
function toCalendarDate(d: Date): CalendarDate {
  return {
    year: d.getUTCFullYear(),
    month: d.getUTCMonth() + 1,
    day: d.getUTCDate(),
  };
}

/** Set a UTC Date to midnight UTC. */
function startOfDayUtc(d: Date): Date {
  return new Date(Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDate()));
}

/** Get confirmed bookings for a user as busy intervals. */
async function getDbBusy(
  userId: number,
  start: Date,
  end: Date,
): Promise<BusyInterval[]> {
  const db = getDb();
  const rows = await db
    .select({ startAt: bookings.startAt, endAt: bookings.endAt })
    .from(bookings)
    .innerJoin(eventTypes, eq(bookings.eventTypeId, eventTypes.id))
    .where(
      and(
        eq(eventTypes.userId, userId),
        eq(bookings.status, "confirmed"),
        lt(bookings.startAt, end),
        gte(bookings.endAt, start),
      ),
    );
  return rows.map((r): BusyInterval => [r.startAt, r.endAt]);
}

/** Load a user + event type or throw. */
async function loadUserAndEventType(
  slug: string,
  eventTypeSlug: string,
): Promise<{
  user: typeof users.$inferSelect;
  eventType: typeof eventTypes.$inferSelect;
  rules: typeof availabilityRules.$inferSelect[];
}> {
  const db = getDb();

  const user = first(
    await db
      .select()
      .from(users)
      .where(and(eq(users.slug, slug), eq(users.isActive, true)))
      .limit(1),
  );
  if (!user) {
    throw Object.assign(new Error("User not found"), { status: 404 });
  }

  const eventType = first(
    await db
      .select()
      .from(eventTypes)
      .where(
        and(
          eq(eventTypes.userId, user.id),
          eq(eventTypes.slug, eventTypeSlug),
          eq(eventTypes.isActive, true),
        ),
      )
      .limit(1),
  );
  if (!eventType) {
    throw Object.assign(new Error("Event type not found"), { status: 404 });
  }

  const rules = await db
    .select()
    .from(availabilityRules)
    .where(eq(availabilityRules.userId, user.id));

  return { user, eventType, rules };
}

// ---------------------------------------------------------------------------
// Public booking page
// ---------------------------------------------------------------------------

booking.get("/:slug", async (c) => {
  const slug = c.req.param("slug");
  if (!slug) return c.json({ error: "Not found" }, 404);

  const db = getDb();
  const user = first(
    await db
      .select({
        id: users.id,
        name: users.name,
        slug: users.slug,
        timezone: users.timezone,
      })
      .from(users)
      .where(and(eq(users.slug, slug), eq(users.isActive, true)))
      .limit(1),
  );
  if (!user) return c.json({ error: "Booking page not found" }, 404);

  const activeEventTypes = await db
    .select({
      id: eventTypes.id,
      slug: eventTypes.slug,
      title: eventTypes.title,
      description: eventTypes.description,
      durationMinutes: eventTypes.durationMinutes,
      color: eventTypes.color,
      intakeQuestions: eventTypes.intakeQuestions,
    })
    .from(eventTypes)
    .where(and(eq(eventTypes.userId, user.id), eq(eventTypes.isActive, true)))
    .orderBy(eventTypes.id);

  return c.json({ user, eventTypes: activeEventTypes });
});

// ---------------------------------------------------------------------------
// Slot availability API
// ---------------------------------------------------------------------------

booking.get("/:slug/:eventTypeSlug/slots", async (c) => {
  const slug = c.req.param("slug");
  const eventTypeSlug = c.req.param("eventTypeSlug");
  if (!slug || !eventTypeSlug) return c.json({ error: "Not found" }, 404);

  const dateStr = c.req.query("target_date");
  const clientTimezone = c.req.query("client_timezone") ?? "UTC";

  if (!dateStr) {
    return c.json({ error: "target_date is required (YYYY-MM-DD)" }, 400);
  }

  let targetDate: CalendarDate;
  try {
    const [y, m, d] = dateStr.split("-").map(Number);
    if (!y || !m || !d) throw new Error("bad date");
    targetDate = { year: y, month: m, day: d };
  } catch {
    return c.json({ error: "Invalid target_date" }, 400);
  }

  let context: Awaited<ReturnType<typeof loadUserAndEventType>>;
  try {
    context = await loadUserAndEventType(slug, eventTypeSlug);
  } catch (e: unknown) {
    const err = e as { status?: number; message?: string };
    return c.json({ error: err.message ?? "Not found" }, (err.status as 404) ?? 404);
  }

  const { user, eventType, rules } = context;

  // Build the day window in UTC for free/busy lookup
  const dayJs = new Date(Date.UTC(targetDate.year, targetDate.month - 1, targetDate.day));
  const dayStart = dayJs;
  const dayEnd = new Date(dayJs.getTime() + 24 * 60 * 60 * 1000);

  const [calBusy, dbBusy] = await Promise.all([
    _calendarProvider.getFreeBusy(user.id, dayStart, dayEnd),
    getDbBusy(user.id, dayStart, dayEnd),
  ]);
  const allBusy = [...calBusy, ...dbBusy];

  const slots = generateSlotsForDate(
    targetDate,
    rules,
    allBusy,
    eventType,
    user.timezone,
  );

  const slotData = slots.map((slotUtc) => {
    const slotEndUtc = new Date(slotUtc.getTime() + eventType.durationMinutes * 60 * 1000);
    const warning = detectTimezoneWarning(slotUtc, slotEndUtc, user.timezone, clientTimezone);

    // Format local time using Intl
    const localStr = new Intl.DateTimeFormat("en-US", {
      timeZone: clientTimezone,
      hour: "2-digit",
      minute: "2-digit",
      hour12: true,
    }).format(slotUtc);

    return {
      utc: slotUtc.toISOString(),
      local: localStr,
      warning,
    };
  });

  return c.json({ date: dateStr, slots: slotData });
});

// ---------------------------------------------------------------------------
// Booking submission
// ---------------------------------------------------------------------------

const bookingRequestSchema = z.object({
  slotUtc: z.string().datetime(),
  clientName: z.string().min(1),
  clientEmail: z.string().email(),
  clientTimezone: z.string().default("UTC"),
  intakeAnswers: z.record(z.unknown()).default({}),
});

booking.post(
  "/:slug/:eventTypeSlug/book",
  zValidator("json", bookingRequestSchema),
  async (c) => {
    const slug = c.req.param("slug");
    const eventTypeSlug = c.req.param("eventTypeSlug");
    if (!slug || !eventTypeSlug) return c.json({ error: "Not found" }, 404);

    const data = c.req.valid("json");

    let context: Awaited<ReturnType<typeof loadUserAndEventType>>;
    try {
      context = await loadUserAndEventType(slug, eventTypeSlug);
    } catch (e: unknown) {
      const err = e as { status?: number; message?: string };
      return c.json({ error: err.message ?? "Not found" }, (err.status as 404) ?? 404);
    }

    const { user, eventType } = context;

    const startUtc = new Date(data.slotUtc);
    const endUtc = new Date(startUtc.getTime() + eventType.durationMinutes * 60 * 1000);

    // Double-check slot availability
    const dayStart = startOfDayUtc(startUtc);
    const dayEnd = new Date(dayStart.getTime() + 24 * 60 * 60 * 1000);

    const [calBusy, dbBusy] = await Promise.all([
      _calendarProvider.getFreeBusy(user.id, dayStart, dayEnd),
      getDbBusy(user.id, dayStart, dayEnd),
    ]);

    const slotEndWithBuffer = new Date(
      endUtc.getTime() + eventType.bufferMinutes * 60 * 1000,
    );
    if (overlapsAnyBusy(startUtc, slotEndWithBuffer, [...calBusy, ...dbBusy])) {
      return c.json({ error: "Slot no longer available" }, 409);
    }

    // Insert booking
    const db = getDb();
    const cancelToken = generateToken();
    const rescheduleToken = generateToken();

    const newBooking = first(
      await db
        .insert(bookings)
        .values({
          eventTypeId: eventType.id,
          clientEmail: data.clientEmail,
          clientName: data.clientName,
          clientTimezone: data.clientTimezone,
          clientData: data.intakeAnswers as Record<string, unknown>,
          startAt: startUtc,
          endAt: endUtc,
          status: "confirmed",
          cancelToken,
          rescheduleToken,
        })
        .returning(),
    );

    if (!newBooking) {
      return c.json({ error: "Failed to create booking" }, 500);
    }

    // Build URLs
    const cfg = getConfig();
    const baseUrl = cfg.BASE_URL.replace(/\/$/, "");
    const cancelUrl = `${baseUrl}/bookings/${cancelToken}/cancel`;
    const rescheduleUrl = `${baseUrl}/bookings/${rescheduleToken}/reschedule`;

    // Try to create calendar event
    const description = [
      `Booking with ${data.clientName}`,
      "",
      ...Object.entries(data.intakeAnswers).map(([k, v]) => `${k}: ${v}`),
    ].join("\n");

    let googleEventId: string | null = null;
    try {
      googleEventId = await _calendarProvider.createEvent(
        user.id,
        `${eventType.title} — ${data.clientName}`,
        startUtc,
        endUtc,
        {
          description,
          attendeeEmail: data.clientEmail,
          createMeetLink: true,
        },
      );
      await db
        .update(bookings)
        .set({ googleEventId })
        .where(eq(bookings.id, newBooking.id));
    } catch {
      // Non-fatal: proceed without calendar event
    }

    // Generate ICS
    const icsContent = generateIcs({
      uid: `booking-${newBooking.id}@bisque-booking`,
      title: `${eventType.title} with ${user.name}`,
      start: startUtc,
      end: endUtc,
      organizerEmail: user.email,
      organizerName: user.name,
      attendeeEmail: data.clientEmail,
      attendeeName: data.clientName,
      description,
      location: eventType.videoLink ?? eventType.location ?? "",
    });
    const icsBuffer = Buffer.from(icsContent, "utf-8");

    // Send confirmation emails (non-fatal)
    try {
      await _emailProvider.sendConfirmationToClient({
        bookingId: newBooking.id,
        clientEmail: data.clientEmail,
        clientName: data.clientName,
        consultantName: user.name,
        start: startUtc,
        end: endUtc,
        clientTimezone: data.clientTimezone,
        cancelUrl,
        rescheduleUrl,
        videoLink: eventType.videoLink,
        icsContent: icsBuffer,
      });
      await _emailProvider.sendConfirmationToConsultant({
        bookingId: newBooking.id,
        consultantEmail: user.email,
        consultantName: user.name,
        clientName: data.clientName,
        clientEmail: data.clientEmail,
        clientData: data.intakeAnswers as Record<string, unknown>,
        start: startUtc,
        end: endUtc,
        consultantTimezone: user.timezone,
        cancelUrl,
        icsContent: icsBuffer,
      });
    } catch (e) {
      console.error("Email sending failed:", e);
    }

    // Fire webhook (non-fatal)
    try {
      await _webhookProvider.dispatch("booking.created", {
        bookingId: newBooking.id,
        clientEmail: data.clientEmail,
        startAt: startUtc.toISOString(),
        endAt: endUtc.toISOString(),
        eventType: eventType.slug,
        consultantSlug: user.slug,
      });
    } catch {
      // Non-fatal
    }

    return c.json(
      {
        bookingId: newBooking.id,
        startAt: startUtc.toISOString(),
        endAt: endUtc.toISOString(),
        cancelUrl,
        rescheduleUrl,
      },
      201,
    );
  },
);

// ---------------------------------------------------------------------------
// Cancel
// ---------------------------------------------------------------------------

booking.get("/bookings/:token/cancel", async (c) => {
  const token = c.req.param("token");
  if (!token) return c.json({ error: "Not found" }, 404);

  const db = getDb();
  const bk = first(
    await db
      .select()
      .from(bookings)
      .where(eq(bookings.cancelToken, token))
      .limit(1),
  );

  if (!bk || bk.status === "cancelled") {
    return c.json({ error: "Booking not found or already cancelled" }, 404);
  }

  await db
    .update(bookings)
    .set({ status: "cancelled", updatedAt: new Date() })
    .where(eq(bookings.id, bk.id));

  // Delete calendar event (non-fatal)
  if (bk.googleEventId) {
    const et = first(
      await db
        .select({ userId: eventTypes.userId })
        .from(eventTypes)
        .where(eq(eventTypes.id, bk.eventTypeId))
        .limit(1),
    );
    if (et) {
      try {
        await _calendarProvider.deleteEvent(et.userId, bk.googleEventId);
      } catch {
        // Non-fatal
      }
    }
  }

  // Send cancellation emails (non-fatal)
  try {
    await _emailProvider.sendCancellation({
      bookingId: bk.id,
      recipientEmail: bk.clientEmail,
      recipientName: bk.clientName,
      start: bk.startAt,
      end: bk.endAt,
      recipientTimezone: bk.clientTimezone,
      cancelledBy: "client",
    });
  } catch {
    // Non-fatal
  }

  return c.json({ status: "cancelled" });
});

// ---------------------------------------------------------------------------
// Reschedule
// ---------------------------------------------------------------------------

booking.get("/bookings/:token/reschedule", async (c) => {
  const token = c.req.param("token");
  if (!token) return c.json({ error: "Not found" }, 404);

  const db = getDb();
  const bk = first(
    await db
      .select()
      .from(bookings)
      .where(eq(bookings.rescheduleToken, token))
      .limit(1),
  );

  if (!bk) return c.json({ error: "Booking not found" }, 404);

  const et = first(
    await db
      .select({ userId: eventTypes.userId, slug: eventTypes.slug })
      .from(eventTypes)
      .where(eq(eventTypes.id, bk.eventTypeId))
      .limit(1),
  );
  if (!et) return c.json({ error: "Event type not found" }, 404);

  const user = first(
    await db
      .select({ slug: users.slug })
      .from(users)
      .where(eq(users.id, et.userId))
      .limit(1),
  );
  if (!user) return c.json({ error: "User not found" }, 404);

  await db
    .update(bookings)
    .set({ status: "rescheduled", updatedAt: new Date() })
    .where(eq(bookings.id, bk.id));

  return c.redirect(`/${user.slug}/${et.slug}`);
});

export { booking as bookingRouter };
